![image.png](https://wolk-aftersales.com/neuewebsite/wp-content/uploads/2021/05/Autodoc_Logo_2018.png)

## <p style="background-color:#FDFEFE; font-family:newtimeroman; color:#060108; font-size:200%; text-align:center; border-radius:10px 10px;">Purchase Funnel Project</p>

<a id="toc"></a>

## <p style="background-color:#33475b; font-family:newtimeroman; color:#FFF9ED; font-size:175%; text-align:center; border-radius:10px 10px;">Content</p>

* [Introduction](#0)
* [I. Importing Related Libraries](#1)
* [II. Load Data and Data Review](#2)
* [III. Focus on Columns -- Feature Engineering](#3)
* [End Of the Project](#4)

## <p style="background-color:#33475b; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">Introduction</p>

<a id="0"> </a>

   It is a project to build a custom purchase funnel to improve marketing strategy and better align it with customer behavior. This funnel will be built based on extensive tracking data and will map the customer journey from their first visit to the relevant website, to viewing products or listings, to the moment of purchas

## Aim of the Notebook

   The primary objectives are to identify and analyze each step of this marketing funnel, uncovering detailed insights  at each stage, enabling us to optimize our approach and uncover conversion rates, and offer improvements based on the rates. 

- As a general process, we will first understand the dataset, and then manipulate and prepare the dataset for the calculations we will use in the next steps. 

- In the next step, various calculations will be shown to reveal and interpret customer behavior at each stage. 

- These results need to be presented in a report in order to make a presentation to the whole team. This way the data will be more understandable and easy to interpret for everyone. For this report, the latest version of the dataset will be imported and a dashboard will be created. 

- Finally, we will generate various insights with the help of the SQL tool. 

### Now, let's start building the notebook for our data and funnel analysis. 

## <p style="background-color:#33475b; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">I. Importing Related Libraries</p>

<a id="1"></a>

In [140]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from matplotlib.patches import Patch
from matplotlib.lines import Line2D
import seaborn as sns

import warnings
warnings.filterwarnings("ignore")
warnings.warn("this will not show")

# Figure&Display options
sns.set_style("whitegrid")
plt.rcParams["figure.figsize"] = (10,6)
#pd.set_option('max_colwidth',200)
#pd.set_option('display.max_rows', 1000)
pd.set_option('display.max_columns', None)
pd.set_option('display.float_format', lambda x: '%.3f' % x)

## <p style="background-color:#33475b; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">II. Load Data and Data Review</p>

<a id="2"></a>

In [149]:
df = pd.read_csv("data_set_da_test.csv")

In [150]:
df.head()

event_date                session                   user  \
0  2022-10-08 17:02:41  14274187577460658115s   2006979063809820329u   
1  2022-10-08 17:06:19  14274187577460658115s   2006979063809820329u   
2  2022-10-08 22:19:47   2704204808571844605s   2007646148110679693u   
3  2022-10-08 22:24:30   8970170322512311099s  11839491588321754710u   
4  2022-10-08 21:22:20  16223970371660715740s  11839887495958431209u   

             page_type event_type  product  
0  search_listing_page  page_view        0  
1  search_listing_page  page_view        0  
2         listing_page  page_view        0  
3  search_listing_page  page_view        0  
4         product_page  page_view        0

In [5]:
df.sample(5)

event_date                session                   user  \
414092  2022-10-12 18:52:19  11914947960018664033s   8880184366266217263u   
13682   2022-09-30 19:43:05  18163706171672094151s  10731902290458823799u   
16904   2022-09-30 23:49:26  10131781634862675530s     51976650537349344u   
457303  2022-10-06 10:13:29   5991075035764835210s  12754865886541179551u   
622329  2022-10-10 21:19:25   3980190479791933092s  10770185070123831430u   

                  page_type event_type  product  
414092  search_listing_page  page_view        0  
13682   search_listing_page  page_view        0  
16904          product_page  page_view        0  
457303         product_page  page_view        0  
622329         product_page  page_view        0

In [143]:
df.shape

(637238, 6)

In [144]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 637238 entries, 0 to 637237
Data columns (total 6 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   event_date  637238 non-null  object
 1   session     637238 non-null  object
 2   user        637238 non-null  object
 3   page_type   637238 non-null  object
 4   event_type  637238 non-null  object
 5   product     637238 non-null  int64 
dtypes: int64(1), object(5)
memory usage: 29.2+ MB


In [151]:
df.duplicated().sum()

839

In [152]:
df.drop_duplicates(inplace=True)

In [153]:
df.duplicated().sum()

0

In [1]:
df.describe(include=object)

NameError: name 'df' is not defined

In [23]:
df.describe(include = object)

event_date               session                   user  \
count                637238                637238                 637238   
unique               464539                340443                 288088   
top     2022-10-05 11:57:23  9543669642136985868s  11769744300065907078u   
freq                      8                  1264                   1266   

           page_type event_type  
count         637238     637238  
unique             4          3  
top     product_page  page_view  
freq          282950     612498

In [6]:
df_object = df.select_dtypes(include='object')

In [7]:
for col in df_object:
    print(f"{col:<20}:", df[col].nunique())

event_date          : 464539
session             : 340443
user                : 288088
page_type           : 4
event_type          : 3


In [24]:
df.isna().sum()

event_date    0
session       0
user          0
page_type     0
event_type    0
product       0
dtype: int64

In [13]:
df.columns

Index(['event_date', 'session', 'user', 'page_type', 'event_type', 'product'], dtype='object')

In [16]:
df.session.value_counts()

session
9543669642136985868s     1258
14816963559813395593s     346
3509679068011838002s      251
7160321411091116049s      203
6410548667176451624s      193
                         ... 
7187446511484344046s        1
15172055876479169932s       1
8304886803886160144s        1
11149641400322386010s       1
8776519620226077650s        1
Name: count, Length: 340443, dtype: int64

In [17]:
df.user.value_counts()

user
11769744300065907078u    1260
10671157392604114939u    1014
15055679654066929981u     977
9697135938868880738u      974
1316617677830457669u      666
                         ... 
9102072801364594820u        1
9103796325275164355u        1
9104387786437595441u        1
15997284830081135375u       1
18416278413345382629u       1
Name: count, Length: 288088, dtype: int64

## <p style="background-color:#33475b; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">III. Focus on Columns -- Feature Engineering</p>

<a id="3"></a>

## event_date

In [80]:
df

event_date                session                   user  \
0       2022-10-08 17:02:41  14274187577460658115s   2006979063809820329u   
1       2022-10-08 17:06:19  14274187577460658115s   2006979063809820329u   
2       2022-10-08 22:19:47   2704204808571844605s   2007646148110679693u   
3       2022-10-08 22:24:30   8970170322512311099s  11839491588321754710u   
4       2022-10-08 21:22:20  16223970371660715740s  11839887495958431209u   
...                     ...                    ...                    ...   
637233  2022-09-30 19:10:47   4337671502003974886s  16408992711249841548u   
637234  2022-09-30 14:44:29  13436616217932629112s  18362900391208328063u   
637235  2022-09-30 02:20:22   6129704733313384581s  18383133457867141415u   
637236  2022-09-30 12:20:04  14145319343967194043s  18403471609558160062u   
637237  2022-09-30 08:10:58   8776519620226077650s  18416278413345382629u   

                  page_type event_type  product  
0       search_listing_page  page_view        0  
1       search_listing_page  page_view        0  
2              listing_page  page_view        0  
3       search_listing_page  page_view        0  
4              product_page  page_view        0  
...                     ...        ...      ...  
637233         listing_page  page_view        0  
637234         listing_page  page_view        0  
637235         listing_page  page_view        0  
637236         listing_page  page_view        0  
637237         listing_page  page_view        0  

[636399 rows x 6 columns]

In [154]:
df[["event_day", "event_hour"]] = df["event_date"].str.split(' ', expand=True)

In [155]:
columns = ['event_date', 'event_day', 'event_hour', 'session', 'user', 'page_type', 'event_type', 'product']

df = df[columns]

In [156]:
df.drop(columns=['event_date'], inplace=True)

In [157]:
df

event_day event_hour                session                   user  \
0       2022-10-08   17:02:41  14274187577460658115s   2006979063809820329u   
1       2022-10-08   17:06:19  14274187577460658115s   2006979063809820329u   
2       2022-10-08   22:19:47   2704204808571844605s   2007646148110679693u   
3       2022-10-08   22:24:30   8970170322512311099s  11839491588321754710u   
4       2022-10-08   21:22:20  16223970371660715740s  11839887495958431209u   
...            ...        ...                    ...                    ...   
637233  2022-09-30   19:10:47   4337671502003974886s  16408992711249841548u   
637234  2022-09-30   14:44:29  13436616217932629112s  18362900391208328063u   
637235  2022-09-30   02:20:22   6129704733313384581s  18383133457867141415u   
637236  2022-09-30   12:20:04  14145319343967194043s  18403471609558160062u   
637237  2022-09-30   08:10:58   8776519620226077650s  18416278413345382629u   

                  page_type event_type  product  
0       search_listing_page  page_view        0  
1       search_listing_page  page_view        0  
2              listing_page  page_view        0  
3       search_listing_page  page_view        0  
4              product_page  page_view        0  
...                     ...        ...      ...  
637233         listing_page  page_view        0  
637234         listing_page  page_view        0  
637235         listing_page  page_view        0  
637236         listing_page  page_view        0  
637237         listing_page  page_view        0  

[636399 rows x 7 columns]

In [85]:
df.event_day.value_counts()

event_day
2022-10-10    52715
2022-10-11    52079
2022-10-12    49702
2022-10-13    49365
2022-10-03    48695
2022-10-04    48144
2022-10-05    47995
2022-09-30    46224
2022-10-06    43753
2022-10-02    43202
2022-10-01    42546
2022-10-09    39407
2022-10-07    36840
2022-10-08    35732
Name: count, dtype: int64

In [63]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 636399 entries, 0 to 637237
Data columns (total 7 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   event_day   636399 non-null  object
 1   event_hour  636399 non-null  object
 2   session     636399 non-null  object
 3   user        636399 non-null  object
 4   page_type   636399 non-null  object
 5   event_type  636399 non-null  object
 6   product     636399 non-null  int64 
dtypes: int64(1), object(6)
memory usage: 38.8+ MB


## event_hour_period

In [158]:
def categorize_time(hour_str):
    
    hour = pd.to_datetime(hour_str, format = '%H:%M:%S').time()
    
    if hour >= pd.to_datetime('00:00:00', format='%H:%M:%S').time() and hour <= pd.to_datetime('08:59:59', format='%H:%M:%S').time():
        return 'morning'
    
    elif hour >= pd.to_datetime('08:59:59', format='%H:%M:%S').time() and hour <= pd.to_datetime('11:59:59', format='%H:%M:%S').time():
        return 'noon'
    
    elif hour >= pd.to_datetime('12:00:00', format='%H:%M:%S').time() and hour <= pd.to_datetime('14:59:59', format='%H:%M:%S').time():
        return 'afternoon'
    
    elif hour >= pd.to_datetime('15:00:00', format='%H:%M:%S').time() and hour <= pd.to_datetime('17:59:59', format='%H:%M:%S').time():
        return 'late afternoon'
    
    elif hour >= pd.to_datetime('18:00:00', format='%H:%M:%S').time() and hour <= pd.to_datetime('20:59:59', format='%H:%M:%S').time():
        return 'evening'
    
    elif hour >= pd.to_datetime('21:00:00', format='%H:%M:%S').time() and hour <= pd.to_datetime('23:59:59', format='%H:%M:%S').time():
        return 'midnight'
    
    else:
        return 'other'

In [159]:
df["event_hour_period"] = df["event_hour"].apply(categorize_time)

In [160]:
df["event_day"] = pd.to_datetime(df["event_day"])

In [161]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 636399 entries, 0 to 637237
Data columns (total 8 columns):
 #   Column             Non-Null Count   Dtype         
---  ------             --------------   -----         
 0   event_day          636399 non-null  datetime64[ns]
 1   event_hour         636399 non-null  object        
 2   session            636399 non-null  object        
 3   user               636399 non-null  object        
 4   page_type          636399 non-null  object        
 5   event_type         636399 non-null  object        
 6   product            636399 non-null  int64         
 7   event_hour_period  636399 non-null  object        
dtypes: datetime64[ns](1), int64(1), object(6)
memory usage: 43.7+ MB


In [162]:
columns = ['event_day', 'event_hour', 'event_hour_period', 'session', 'user', 'page_type', 'event_type', 'product']

df = df[columns]

In [163]:
df

event_day event_hour event_hour_period                session  \
0      2022-10-08   17:02:41    late afternoon  14274187577460658115s   
1      2022-10-08   17:06:19    late afternoon  14274187577460658115s   
2      2022-10-08   22:19:47          midnight   2704204808571844605s   
3      2022-10-08   22:24:30          midnight   8970170322512311099s   
4      2022-10-08   21:22:20          midnight  16223970371660715740s   
...           ...        ...               ...                    ...   
637233 2022-09-30   19:10:47           evening   4337671502003974886s   
637234 2022-09-30   14:44:29         afternoon  13436616217932629112s   
637235 2022-09-30   02:20:22           morning   6129704733313384581s   
637236 2022-09-30   12:20:04         afternoon  14145319343967194043s   
637237 2022-09-30   08:10:58           morning   8776519620226077650s   

                         user            page_type event_type  product  
0        2006979063809820329u  search_listing_page  page_view        0  
1        2006979063809820329u  search_listing_page  page_view        0  
2        2007646148110679693u         listing_page  page_view        0  
3       11839491588321754710u  search_listing_page  page_view        0  
4       11839887495958431209u         product_page  page_view        0  
...                       ...                  ...        ...      ...  
637233  16408992711249841548u         listing_page  page_view        0  
637234  18362900391208328063u         listing_page  page_view        0  
637235  18383133457867141415u         listing_page  page_view        0  
637236  18403471609558160062u         listing_page  page_view        0  
637237  18416278413345382629u         listing_page  page_view        0  

[636399 rows x 8 columns]

In [17]:
df.event_hour_period.value_counts()

event_hour_period
late afternoon    123608
afternoon         119963
evening           112915
noon              102144
morning            97928
midnight           79841
Name: count, dtype: int64

## user

In [111]:
df.user.value_counts()

user
11769744300065907078u    1260
10671157392604114939u    1014
15055679654066929981u     977
9697135938868880738u      974
1316617677830457669u      666
                         ... 
9102072801364594820u        1
9103796325275164355u        1
9104387786437595441u        1
15997284830081135375u       1
18416278413345382629u       1
Name: count, Length: 288088, dtype: int64

In [164]:
df.user.value_counts().max()

1260

In [165]:
counts = df.user.value_counts()

In [166]:
count_one = (counts == 1).sum()

print(f'{count_one} users joined the website only one.')

197101 users joined the website only one.


In [167]:
count_two = counts[(counts > 1) & (counts <= 100)]

In [168]:
count_two = count_two.size

print(f'{count_two} users joined the website more than 1 and less 100.')

90933 users joined the website more than 1 and less 100.


In [169]:
count_three = counts[(counts > 100) & (counts <= 500)]

In [170]:
count_three = count_three.size

print(f'{count_three} users joined the website more than 100 and less 500.')

47 users joined the website more than 100 and less 500.


In [171]:
count_four = counts[(counts > 500) & (counts <= 1260)]  # 1260 is max number

In [172]:
count_four

user
11769744300065907078u    1260
10671157392604114939u    1014
15055679654066929981u     977
9697135938868880738u      974
1316617677830457669u      666
3014248931872578225u      532
6732678485673080759u      510
Name: count, dtype: int64

In [26]:
count_four = count_four.size

print(f'{count_four} users joined the website more than 500 and less 1500.')

7 users joined the website more than 500 and less 1500.


In [27]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 636399 entries, 0 to 637237
Data columns (total 8 columns):
 #   Column             Non-Null Count   Dtype 
---  ------             --------------   ----- 
 0   event_day          636399 non-null  object
 1   event_hour         636399 non-null  object
 2   event_hour_period  636399 non-null  object
 3   session            636399 non-null  object
 4   user               636399 non-null  object
 5   page_type          636399 non-null  object
 6   event_type         636399 non-null  object
 7   product            636399 non-null  int64 
dtypes: int64(1), object(7)
memory usage: 43.7+ MB


In [28]:
df_object = df.select_dtypes(include='object')

In [173]:
for col in df_object:
    print(f"{col:<20}:", df[col].nunique())

event_day           : 14
event_hour          : 83250
event_hour_period   : 6
session             : 340443
user                : 288088
page_type           : 4
event_type          : 3


## page_type

In [31]:
df.page_type.value_counts()

page_type
product_page           282714
listing_page           231369
search_listing_page    113576
order_page               8740
Name: count, dtype: int64

## event_type

In [32]:
df.event_type.value_counts()

event_type
page_view      611739
add_to_cart     15920
order            8740
Name: count, dtype: int64

## product

In [114]:
df['product'].value_counts()

product
0           620479
26372760        21
27133309        20
26780006        14
20283956        13
             ...  
27661741         1
34582668         1
19990749         1
26165071         1
35704689         1
Name: count, Length: 12442, dtype: int64

### Most Viewed Products

In [110]:
most_popular_product = df[(df["page_type"] == 'product_page') | (df["page_type"] == 'listing_page') | (df["page_type"] == 'search_listing_page')]['product'].value_counts().reset_index()

In [117]:
most_popular_product.head(6)

product   count
0         0  611739
1  26372760      21
2  27133309      20
3  26780006      14
4  20283956      13
5  20053168      13

The '0' product is mean that users viewed only page so that it is not a product. Thats why the second row shows the most viewed product to us which is the product ID **26372760** with **21** times.

### Most Added to Cart Products

In [115]:
most_added_to_cart_products = df[(df["event_type"] == 'add_to_cart')]['product'].value_counts().reset_index()

In [120]:
most_added_to_cart_products.head(5)

product  count
0  26372760     21
1  27133309     20
2  26780006     14
3  20053168     13
4  20283956     13

After the viewing the first three products, they are most added to chart products as well. But the fourth line is changed although they had same viewes. Its stange situation and it may be a case that needs to be investigated.

### Most Purchased Products

In [125]:
most_purchased_products = df[(df["event_type"] == 'add_to_chart')]['product'].value_counts().reset_index()

In [126]:
most_purchased_products

product  count
0        0   8740

In [133]:
df[(df["event_type"] == 'order') & (df["product"] != 0)]

Empty DataFrame
Columns: [event_day, event_hour, event_hour_period, session, user, page_type, event_type, product]
Index: []

In [177]:
df.to_csv("last_version_dataset.csv")

## <p style="background-color:#FDFEFE; font-family:newtimeroman; color:#060108; font-size:200%; text-align:center; border-radius:10px 10px;">Marketing Funnel Analysis</p>

To uncover insights on a website, every single activity of the customer on the website needs to be examined.  In this way, a marketing funnel will show the customer journey from visiting the website and viewing a product or listing, to initiating the purchase and the moment of purchase. By analyzing each stage, it is aimed to provide an advantage for senior management or business partners to make business decisions. 

To recap these key metrics:

  1. **Awareness**     : Visiting the website by user. 
  
  2. **Interest**      : The user starts researching the product to gather more information about its features in product page.
  
  3. **Consideration** : The prospect compares the product with other products searches from the listing page.
  
  4. **Intent**        : When the potential customer decides to buy the product, but there is a final stage left to make a    commitment.
  
  5. **Purchase**      : Placing an order for the product(s). 
  
  
Next step, should be performed various calculations for the key metrics that have determined and I will reveale the behavior of the customer on the website finally.

## Page Views to Awareness

In [76]:
product_view = df[df["page_type"] == 'product_page']

In [79]:
len(product_view)

282714

In [82]:
listing_view = df[(df["page_type"] == 'listing_page') | (df["page_type"] == 'search_listing_page')]

In [84]:
len(listing_view)

344945

In [94]:
len(product_view['user'].unique())

142515

In [95]:
len(listing_view['user'].unique())

168396

In [85]:
conversion_rate_awareness = len(product_view['user'].unique())/len(listing_view['user'].unique())

In [86]:
conversion_rate_awareness

0.8463087009192617

This value indicates that around **84.63%** of the visitors to the website during the awareness phase performed a desired action. This high conversion rate shows that this first stage on the website effectively engages visitors and converts them into leads or customers.

## Interest (Consideration) Conversion Rate

In [89]:
add_to_chart_actions = df[df["event_type"] == 'add_to_cart']

In [92]:
len(add_to_chart_actions)

15920

In [93]:
len(add_to_chart_actions['user'].unique())

10060

In [90]:
conversion_rate_consideration = len(add_to_chart_actions['user'].unique())/len(listing_view['user'].unique())

In [91]:
conversion_rate_consideration

0.05974013634528136

This value indicates that approximately **5.97%** of the potential customers who evaluated the product prices successfully converted. This means that a small percentage of all users who research and evaluate the product ultimately decide to buy.

## Purchase Conversion Rate

In [96]:
purchase_action = df[df["event_type"] == 'order']

In [97]:
len(purchase_action['user'].unique())

7338

In [98]:
conversion_rate_purchase = len(purchase_action['user'].unique())/len(add_to_chart_actions['user'].unique())

In [99]:
conversion_rate_purchase

0.7294234592445328

Purchase Conversion Rate **72.94%** is a strong indicator of an effective checkout process. This value, which shows the order conversion rate of the products added to the chart, can be interpreted that the payment system is user friendly, fast, and secure.

In [178]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 636399 entries, 0 to 637237
Data columns (total 8 columns):
 #   Column             Non-Null Count   Dtype         
---  ------             --------------   -----         
 0   event_day          636399 non-null  datetime64[ns]
 1   event_hour         636399 non-null  object        
 2   event_hour_period  636399 non-null  object        
 3   session            636399 non-null  object        
 4   user               636399 non-null  object        
 5   page_type          636399 non-null  object        
 6   event_type         636399 non-null  object        
 7   product            636399 non-null  int64         
dtypes: datetime64[ns](1), int64(1), object(6)
memory usage: 43.7+ MB
